File: app/src/main/java/uk/co/victoriajanedavis/chatapp/data/repositories/SentFriendRequestRepository.java
package uk.co.victoriajanedavis.chatapp.data.repositories;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.Single;
import io.reactivex.annotations.NonNull;
import uk.co.victoriajanedavis.chatapp.injection.scopes.ApplicationScope;
import uk.co.victoriajanedavis.chatapp.injection.qualifiers.SentFriendRequestStore;
import uk.co.victoriajanedavis.chatapp.data.mappers.FriendshipDbEntityMapper;
import uk.co.victoriajanedavis.chatapp.data.mappers.UserNwFriendshipDbMapper;
import uk.co.victoriajanedavis.chatapp.data.model.db.FriendshipDbModel;
import uk.co.victoriajanedavis.chatapp.data.repositories.store.BaseReactiveStore;
import uk.co.victoriajanedavis.chatapp.data.services.ChatAppService;
import uk.co.victoriajanedavis.chatapp.domain.entities.FriendshipEntity;

@ApplicationScope
public class SentFriendRequestRepository {

    private final BaseReactiveStore<FriendshipDbModel> friendStore;
    private final ChatAppService chatService;

    private final FriendshipDbEntityMapper dbEntityMapper;
    private final UserNwFriendshipDbMapper nwSentDbMapper;

    @Inject
    public SentFriendRequestRepository(@NonNull @SentFriendRequestStore BaseReactiveStore<FriendshipDbModel> friendStore,
                                       @NonNull ChatAppService service) {
        this.friendStore = friendStore;
        this.chatService = service;

        dbEntityMapper = new FriendshipDbEntityMapper();
        nwSentDbMapper = new UserNwFriendshipDbMapper(false, true);
    }

    @NonNull
    public Observable<List<FriendshipEntity>> getAllSentFriendRequests() {
        return friendStore.getAll(null)
                .switchMapSingle(dbModels -> Observable.fromIterable(dbModels)
                        .map(dbEntityMapper::mapFrom)
                        .toList());
    }

    @NonNull
    public Completable fetchSentFriendRequests() {
        return chatService.getSentFriendRequests()
                .flatMap(nwModels -> Observable.fromIterable(nwModels)
                        .map(nwSentDbMapper::mapFrom)
                    .toList())
                .flatMapCompletable(friendRequests -> friendStore.replaceAll(null, friendRequests));
    }

    @NonNull
    public Single<FriendshipEntity> sendFriendRequest(String username) {
        return chatService.sendFriendRequest(username)
                .map(nwSentDbMapper::mapFrom)
                .flatMap(friendEntity -> friendStore.storeSingular(friendEntity)
                        .andThen(Single.just(friendEntity).map(dbEntityMapper::mapFrom)));
    }

    @NonNull
    public Single<FriendshipEntity> cancelSentFriendRequest(String username) {
        return chatService.cancelSentFriendRequest(username)
                .map(nwSentDbMapper::mapFrom)
                .flatMap(friendEntity -> friendStore.delete(friendEntity)
                        .andThen(Single.just(friendEntity).map(dbEntityMapper::mapFrom)));
    }


}

File: app/src/main/java/uk/co/victoriajanedavis/chatapp/data/repositories/SentFriendRequestRepository.kt
package uk.co.victoriajanedavis.chatapp.data.repositories

import javax.inject.Inject

import io.reactivex.Completable
import io.reactivex.Observable
import uk.co.victoriajanedavis.chatapp.injection.scopes.ApplicationScope
import uk.co.victoriajanedavis.chatapp.data.mappers.FriendshipDbEntityMapper
import uk.co.victoriajanedavis.chatapp.data.mappers.UserNwFriendshipDbMapper
import uk.co.victoriajanedavis.chatapp.data.model.db.FriendshipDbModel
import uk.co.victoriajanedavis.chatapp.data.repositories.store.BaseReactiveStore
import uk.co.victoriajanedavis.chatapp.data.repositories.store.SentFriendRequestStore
import uk.co.victoriajanedavis.chatapp.data.services.ChatAppService
import uk.co.victoriajanedavis.chatapp.domain.entities.FriendshipEntity
import uk.co.victoriajanedavis.chatapp.domain.entities.FriendshipLoadingState.*
import uk.co.victoriajanedavis.chatapp.domain.common.doOnErrorOrDispose
import uk.co.victoriajanedavis.chatapp.domain.common.mapList
import java.util.UUID
import javax.inject.Named

@ApplicationScope
class SentFriendRequestRepository @Inject constructor(
    @Named(SentFriendRequestStore) private val friendStore: BaseReactiveStore<FriendshipDbModel>,
    private val chatService: ChatAppService
) {
    private val dbEntityMapper = FriendshipDbEntityMapper()
    private val nwSentDbMapper = UserNwFriendshipDbMapper(false, true)

    fun getAllSentFriendRequests(): Observable<List<FriendshipEntity>> {
        return friendStore.getAll(null)
            .mapList(dbEntityMapper::mapFrom)
    }

    fun fetchSentFriendRequests(): Completable {
        return chatService.sentFriendRequests
            .mapList(nwSentDbMapper::mapFrom)
            .flatMapCompletable { friendRequests -> friendStore.replaceAll(null, friendRequests) }
    }

    fun sendFriendRequest(username: String, message: String?): Completable {
        return chatService.sendFriendRequest(username)
            .map(nwSentDbMapper::mapFrom)
            .flatMapCompletable(friendStore::storeSingular)
    }

    fun cancelSentFriendRequest(receiverUserUuid: UUID): Completable {
        return friendStore.getSingular(receiverUserUuid)
            .switchMapSingle { friendDbModel -> friendStore.storeSingular(friendDbModel.copy(loadingState = CANCEL))
                .andThen(chatService.cancelSentFriendRequest(friendDbModel.username)
                    .doOnErrorOrDispose {
                        friendStore.storeSingular(friendDbModel.copy(loadingState = NONE)).blockingAwait()
                    })
            }
            .map(nwSentDbMapper::mapFrom)
            .flatMapCompletable(friendStore::delete)
    }
}
